File: tools/gencode.py
#!/bin/env python3

code = {}
buf = []
event = 0

def flush_code():
	global buf
	if event > 0:
		code[event] = buf
		buf = []

event = 0
for line in open("events.txt").readlines():
	line = line.strip()
	if line == "EOF":
		break
	elif line == "":
		continue
	elif line.startswith("#"):
		continue
	elif line.startswith("EVENT"):
		flush_code()
		event = (int(line.split()[1]) << 1)
	elif line.startswith("SHADED"):
		flush_code()
		event = (int(line.split()[1]) << 1) + 1
	else:
		buf.append(line)

flush_code()

code_index = ["-1"] * (72 * 2)

pc = 0
print("const CODE = [")
for event in range(2,146):
	if event & 1 == 1:
		print("\t// SHADED", event >> 1)
	else:
		print("\t// EVENT", event >> 1)
	if not event in code:
		print("\t// TODO")
		continue
	code_index[event-2] = str(pc)
	for line in code[event]:
		if line.startswith('space'):
			line = line.split(' ', 2)
			print('\t[ vm_space, ' + line[1] + ', (s)=>' + line[2] + ' ],')
		elif line.startswith('each space'):
			line = line.split(' ', 2)
			print('\t[ vm_space, 0, (s)=>' + line[2] + ' ],')
		elif line.startswith('piece'):
			line = line.split(' ', 2)
			print('\t[ vm_piece, ' + line[1] + ', (p,s)=>' + line[2] + ' ],')
		elif line.startswith('each piece'):
			line = line.split(' ', 2)
			print('\t[ vm_piece, 0, (p,s)=>' + line[2] + ' ],')
		elif line.startswith('while'):
			line = line.split(' ', 1)
			print('\t[ vm_while, ()=>' + line[1] + ' ],')
		elif line.startswith('if'):
			line = line.split(' ', 1)
			print('\t[ vm_if, ()=>' + line[1] + ' ],')
		elif line.startswith('prompt'):
			line = line.split(' ', 1)
			print('\t[ vm_prompt, "' + line[1] + '" ],')
		elif line.startswith('log'):
			line = line.split(' ', 1)
			print('\t[ vm_log, "' + line[1] + '" ],')
		else:
			line = line.split(' ')
			cmd = line[0]
			args = [ ("()=>" + x if x[0] == '(' else x) for x in line[1:] ]
			if len(args) > 0:
				print('\t[ vm_' + cmd + ', ' + ', '.join(args) + ' ],')
			else:
				print('\t[ vm_' + cmd + ' ],')
		pc = pc + 1
	print('\t[ vm_return ],')
	pc = pc + 1
print("]")
print("const CODE_INDEX = [ " + ", ".join(code_index) + " ]")

File: tools/gencode.js
"use strict"

let fs = require("fs")

let pc = 0
let UCODE = Array(72).fill(0)
let SCODE = Array(72).fill(0)

function emit(line) {
	++pc
	line[0] = "vm_" + line[0]
	for (let i = 1; i < line.length; ++i) {
		if (typeof line[i] === "string") {
			if (line[i] === "all")
				line[i] = 999
			if (line[i][0] === "(" && !line[i].match(/\)=>/))
				line[i] = "()=>" + line[i]
			if (line[i][0] === "`")
				line[i] = "()=>" + line[i]
		}
	}
	console.log("\t[ " + line.join(", ") + " ],")
}

console.log("const CODE = []")
let first = false

for (let line of fs.readFileSync("events.txt", "utf-8").split("\n")) {
	line = line.trim()
	if (line.length === 0 || line[0] === "#")
		continue
	if (line === "EOF")
		break
	line = line.split(" ")
	switch (line[0]) {
	case "EVENT":
		if (first++) {
			emit(["return"])
			console.log("]")
		}
		UCODE[line[1]] = pc
		console.log("")
		console.log("// EVENT " + line[1])
		console.log("CODE[" + line[1] + " * 2 + 0] = [")
		break
	case "SHADED":
		if (first++) {
			emit(["return"])
			console.log("]")
		}
		SCODE[line[1]] = pc
		console.log("")
		console.log("// SHADED " + line[1])
		console.log("CODE[" + line[1] + " * 2 + 1] = [")
		break

	case "if_space":
		emit([ "if", "()=>can_vm_space(1,(s)=>" + line.slice(1).join(" ") + ")" ])
		emit([ "space", true, 1, 1, "(s)=>" + line.slice(1).join(" ") ])
		emit([ "else" ])
		break

	case "or_space":
		emit([ "space", true, 1, 1, "(s)=>" + line.slice(1).join(" ") ])
		emit([ "endif" ])
		break

	case "space_no_undo":
		emit([ "space", false, line[1], line[1], "(s)=>" + line.slice(2).join(" ") ])
		break
	case "space":
		emit([ "space", true, line[1], line[1], "(s)=>" + line.slice(2).join(" ") ])
		break
	case "space_opt":
		emit([ "space", true, 0, line[1], "(s)=>" + line.slice(2).join(" ") ])
		break

	case "piece":
		emit([ "piece", false, line[1], line[1], "(p,s)=>" + line.slice(2).join(" ") ])
		break
	case "piece_undo":
		emit([ "piece", true, line[1], line[1], "(p,s)=>" + line.slice(2).join(" ") ])
		break
	case "piece_range":
		emit([ "piece", false, line[1], line[2], "(p,s)=>" + line.slice(3).join(" ") ])
		break
	case "piece_opt":
		emit([ "piece", false, 0, line[1], "(p,s)=>" + line.slice(2).join(" ") ])
		break
	case "piece_undo_opt":
		emit([ "piece", true, 0, line[1], "(p,s)=>" + line.slice(2).join(" ") ])
		break

	case "shipment":
		emit([ "shipment", false, line[1], line[1], "(p,s)=>" + line.slice(2).join(" ") ])
		break
	case "shipment_opt":
		emit([ "shipment", false, 0, line[1], "(p,s)=>" + line.slice(2).join(" ") ])
		break

	case "place":
		emit([ "place", false, 0, line[1], line[2] ])
		break
	case "place_opt":
		emit([ "place", false, 1, line[1], line[2] ])
		break
	case "auto_place":
		emit([ "auto_place", false, 0, line[1], line[2] ])
		break

	case "log":
	case "prompt":
		emit([ line[0], line.slice(1).join(" ") ])
		break

	case "count_spaces":
		emit([ line[0], "(s)=>" + line.slice(1).join(" ") ])
		break
	case "count_pieces":
		emit([ line[0], "(p,s)=>" + line.slice(1).join(" ") ])
		break

	case "asm":
	case "if":
	case "while":
		emit([ line[0], "()=>" + line.slice(1).join(" ") ])
		break

	default:
		emit(line)
		break
	}
}

emit(["return"])
console.log("]")
